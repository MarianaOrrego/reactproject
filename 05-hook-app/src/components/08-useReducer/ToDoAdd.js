import { useForm } from "../../hooks/useForm";

export const ToDoAdd = ({ handleAddToDo }) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {

        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }
        const nuevaTarea = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddToDo( nuevaTarea );
        reset();
    }

    return (
        <>
            <h4>Agregar tarea</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    autoComplete='off'
                    className='form-control'
                    name='description'
                    onChange={handleInputChange}
                    placeholder='Tarea pendiente...'
                    type='text'
                    value={description}
                />

                <div className='d-flex justify-content-center mt-3'>
                    <button className='btn btn-outline-primary' type='submit'>
                        Agregar
                    </button>
                </div>

            </form>
        </>
    )
}

