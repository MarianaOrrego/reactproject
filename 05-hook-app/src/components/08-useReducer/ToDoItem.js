export const ToDoItem = ({ tarea, i, handleDelete, handleToggle}) => {
    return (
        <li
            key={tarea.id}
            className='list-group-item'
        >
            <p
                className={` ${tarea.done && 'complete'} `}
                onClick={() => handleToggle(tarea.id)}
            >
                {i + 1}.  {tarea.desc}
            </p>

            <button
                className='btn btn-danger'
                onClick={() => handleDelete(tarea.id)}
            >
                Borrar
            </button>
        </li>
    )
}
