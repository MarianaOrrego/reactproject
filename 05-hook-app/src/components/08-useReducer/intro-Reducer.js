const initialState = [{
    id: 1, 
    tarea: 'Comprar chocolate',
    done: false
}];

const tareaReducer = ( state = initialState, action ) => {

    if (action?.type === 'agregar'){
        return [ ...state, action.payload];
    }

    return state;
}

let tarea = tareaReducer();

const nuevaTarea = {
    id: 2, 
    tarea: 'Comprar leche',
    done: false
}

const agregarAction ={
    type: 'agregar',
    payload: nuevaTarea
}

tarea = tareaReducer( tarea, agregarAction)

console.log( tarea )