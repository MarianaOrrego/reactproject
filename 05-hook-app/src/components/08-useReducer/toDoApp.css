body {
    padding: 50px;
}

h1 {
    text-align: center;
}

h4 {
    text-align: center;
}

p{
    cursor: pointer;
    margin: 0px;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through red;
}