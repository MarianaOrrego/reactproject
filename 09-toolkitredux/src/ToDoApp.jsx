import { useState } from "react";
import { useGetToDosQuery, useGetToDosIdQuery } from "./store/apis/toDosApi"

export const ToDoApp = () => {

    const [todoId, setTodoId] = useState(1);

    const {data: todo, isLoading} = useGetToDosIdQuery(todoId);

    const nextToDo = () => {
        setTodoId( todoId + 1);
    }

    const prevToDo = () => {
        if(todoId === 1) return;
        setTodoId( todoId - 1);
    }

  return (
    <>
      <h1>To Do - RTK Query</h1>
      <hr/>
      <h4>isLoading: {isLoading ? 'True': 'False'}</h4>

      <pre>{ JSON.stringify(todo)}</pre>

      <button onClick={prevToDo}>
        Anterior tarea
      </button>
      <button onClick={nextToDo}>
        Siguiente tarea
      </button>

      {/* <ul>
        {todos.map (toDo => (
            <li key={toDo.id}>
                <strong>{ toDo.completed ? 'Completo ' : 'Pendiente ' }</strong>
                { toDo.title }
            </li>
        ))}
      </ul> */}

      
    </>
  )
}