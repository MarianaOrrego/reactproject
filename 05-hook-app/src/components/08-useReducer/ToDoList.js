import PropTypes from 'prop-types';
import { ToDoItem } from "./ToDoItem"

export const ToDoList = ( { tareas, handleDelete, handleToggle } ) => {
    return (
        <ul className='list-group list-group-flush'>
            {
                tareas.map((tarea, i) => (
                    <ToDoItem
                        key={ tarea.id } 
                        tarea= { tarea } 
                        i = { i }
                        handleDelete = { handleDelete } 
                        handleToggle = { handleToggle }
                    />
                ))
            }
        </ul>
    )
}

ToDoList.propTypes = {
    tareas: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}