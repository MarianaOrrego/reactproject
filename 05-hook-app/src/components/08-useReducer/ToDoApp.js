import { useEffect, useReducer } from 'react';
import { toDoReducer } from './toDoReducer';
import './toDoApp.css';
import { ToDoList } from './ToDoList';
import { ToDoAdd } from './ToDoAdd';

const init = () => {

  return JSON.parse(localStorage.getItem('Tareas')) || [];

}

export const ToDoApp = () => {

  const [tareas, dispatch] = useReducer(toDoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('Tareas', JSON.stringify(tareas));
  }, [tareas]);

  const handleDelete = (tareaId) => {

    const action = {
      type: 'delete',
      payload: tareaId
    }

    dispatch(action);
  }

  const handleToggle = (tareaId) => {

    dispatch({
      type: 'toggle',
      payload: tareaId
    })

  }

  const handleAddToDo = ( nuevaTarea ) =>{

    dispatch({
      type: 'add',
      payload: nuevaTarea
    })

  }


  return (
    <div>
      <h1>To do list App ( { tareas.length } )</h1>
      <hr />

      <div className='row'>
        <div className='col-md-6'>

          <ToDoList 
            tareas = { tareas }
            handleDelete = { handleDelete }
            handleToggle = { handleToggle }
          />

        </div>
        <div className='col-md-6'>

          <ToDoAdd 
            handleAddToDo = { handleAddToDo }
          />

        </div>
      </div>

    </div>
  )
}

