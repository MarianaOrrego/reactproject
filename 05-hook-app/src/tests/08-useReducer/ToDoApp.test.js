import { mount, shallow } from 'enzyme';
import { ToDoApp } from '../../components/08-useReducer/ToDoApp';
import { demoToDo } from '../fixtures/demoToDo';
import { act } from '@testing-library/react';

describe('Pruebas en <ToDoApp />', () => {

    const wrapper = shallow( <ToDoApp />);
    Storage.prototype.setItem = jest.fn( () =>{});
  
    test('debe de mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();
      
    });

    test('debe de agregar un ToDo', () => {

        const wrapper = mount( <ToDoApp /> );

        act( () => {
            wrapper.find('ToDoAdd').prop('handleAddToDo')( demoToDo[0] );
            wrapper.find('ToDoAdd').prop('handleAddToDo')( demoToDo[1] );
        });

        expect( wrapper.find('h1').text().trim() ).toBe('To do list App ( 2 )');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
        
    });

    test('debe eliminar un ToDo', () => {
      
        wrapper.find('ToDoAdd').prop('handleAddToDo')(demoToDo[0] );
        wrapper.find('ToDoList').prop('handleDelete')(demoToDo[0].id );

        expect( wrapper.find('h1').text().trim()).toBe('To do list App ( 0 )');

    });
    
})
