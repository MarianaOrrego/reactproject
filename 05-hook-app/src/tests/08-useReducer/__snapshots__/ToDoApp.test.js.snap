// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ToDoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    To do list App ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <ToDoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        tareas={Array []}
      />
    </div>
    <div
      className="col-md-6"
    >
      <ToDoAdd
        handleAddToDo={[Function]}
      />
    </div>
  </div>
</div>
`;
